pysteps/utils/interpolate.py

```diff
@@ -42,186 +41,134 @@ def rbfinterp2d(
-
-    input_array = np.copy(input_array)
-
-    if np.any(~np.isfinite(input_array)):
-        raise ValueError("input_array contains non-finite values")
-
     if input_array.ndim == 1:
         nvar = 1
         input_array = input_array[:, None]
 
     elif input_array.ndim == 2:
         nvar = input_array.shape[1]
 
-    else:
-        raise ValueError(
-            "input_array must have 1 (n) or 2 dimensions (n, m), but it has %i"
-            % input_array.ndim
-        )
-
     npoints = input_array.shape[0]
 
-    if npoints == 0:
-        raise ValueError(
-            "input_array (n, m) must contain at least one sample, but it has %i"
-            % npoints
-        )
-
-    # only one sample, return uniform fields
-    elif npoints == 1:
-        output_array = np.ones((nvar, ygrid.size, xgrid.size))
-        for i in range(nvar):
-            output_array[i, :, :] *= input_array[:, i]
-        return output_array
-
-    coord = np.copy(coord)
-
-    if coord.ndim != 2:
-        raise ValueError(
-            "coord must have 2 dimensions (n, 2), but it has %i" % coord.ndim
-        )
-
-    if npoints != coord.shape[0]:
-        raise ValueError(
-            "the number of samples in the input_array does not match the "
-            + "number of coordinates %i!=%i" % (npoints, coord.shape[0])
-        )
-
-    # normalize coordinates
-    qcoord = np.percentile(coord, [2, 98], axis=0)
-    dextent = np.max(np.diff(qcoord, axis=0))
-    coord = (coord - qcoord[0, :]) / dextent
-
-    rbfunction = rbfunction.lower()
-    if rbfunction not in _rbfunctions:
-        raise ValueError(
-            "Unknown rbfunction '{}'\n".format(rbfunction)
-            + "The available rbfunctions are: "
-            + str(_rbfunctions)
-        ) from None
-
     # generate the target grid
-    X, Y = np.meshgrid(xgrid, ygrid)
-    grid = np.column_stack((X.ravel(), Y.ravel()))
-    # normalize the grid coordinates
-    grid = (grid - qcoord[0, :]) / dextent
+    xgridv, ygridv = np.meshgrid(xgrid, ygrid)
+    gridv = np.column_stack((xgridv.ravel(), ygridv.ravel()))
+
+    # grid mean resolution
+    x_res = np.gradient(xgrid)
+    y_res = np.gradient(ygrid)
+    mean_res = np.mean(np.abs([x_res.mean(), y_res.mean()]))
 
-    # k-nearest interpolation
-    if k is not None and k > 0:
+    if k is not None:
         k = int(np.min((k, npoints)))
+        tree = _cKDTree_cached(coord, hkey=kwargs.get("hkey", None))
+        dist, inds = tree.query(gridv, k=k)
+        if dist.ndim == 1:
+            dist = dist[..., None]
+            inds = inds[..., None]
+    else:
+        # use all points
+        dist = cdist(coord, gridv, "euclidean").transpose()
+        inds = np.arange(npoints)[None, :] * np.ones((gridv.shape[0], npoints)).astype(
+            int
+        )
 
-        # create cKDTree object to represent source grid
-        tree = scipy.spatial.cKDTree(coord)
+    # compute distance-based weights
+    dist = dist / mean_res + 0.5

```

Comment:
I do not understand this line. Why do you divide the distance by the mean resolution?
Should lines 74-77 be moved jyst before this line?

Reply:
The idea was to express distances in terms of pixels rather than geographical units. This way, I could simply add 0.5 (half a pixel) to represent the fact that we are interpolating on a grid rather than to points (the distance can't never be 0). But I'm less and less conviced about this.... should I just remove this part (including lines 74-77)?

Reply:
Ok. I think we can leave it in and just add a more informative comment before line 94. 
If you move lines 74-77 just before we have a clear block if code dealing with the issue. 
I like the idea of not having zero distance. It somehow adds a mini nugget effect to IDW. In case of outliers the interpolated field would be slightly less wild. Btw, is there a way to add an-priori noise term to IDW?

Reply:
Great, thanks a lot for the feedback! Done in 29b52dd

Reply:
Should we make the constant 0.5 a user-selectable parameter?

Reply:
uhm I guess it wouldn't hurt, although I don't think this represents a very sensitive parameter. But might be anyway more transparent to have it listed as tunable parameter.

Reply:
done in a5a1a44